### PR TITLE
Fix listener service

### PR DESCRIPTION
**Pull request by mustafaiman**
Previously, listener service did not register listeners to all members. It used only one member. Also, when that member died, it did not re-register the listeners to another member. Therefore, listeners are lost after reconnection.

This PR changes ListenerService so that

- all listeners are registered to all members and they receive local-only events.
- when a new connection is established, all existing listeners are registered to the new member
- client periodically verifies connection status to all known members
- listener service internal api is changed
- all proxies implement new internal api

- PR also includes a small fix in invocation service that caused listener service to brake. When ownerconnection is null, invocationservice retries the operation.

~~One test does not pass because of a bug in partition service. This PR will be green after merging https://github.com/hazelcast/hazelcast-nodejs-client/pull/194 and rebasing.~~

Fixes https://github.com/hazelcast/hazelcast-nodejs-client/issues/167

### CODE REVIEW

**Comment by mustafaiman** [comment]
verify

File: src/invocation/InvocationService.ts [hunk 1]
```diff
@@ -322,43 +351,221 @@ export class InvocationService {
 /**
  * Handles registration and de-registration of cluster-wide events listeners.
  */
-export class ListenerService {
+export class ListenerService implements ConnectionHeartbeatListener {
```
Comment (asimarslan) [inline src/invocation/InvocationService.ts]: Can we move this class to its own file ?

File: src/invocation/InvocationService.ts [hunk 2]
```diff
@@ -322,43 +351,221 @@ export class InvocationService {
...
+        this.activeRegistrations.forEach((registrationsOnUserKey: Map<ClientConnection, ClientEventRegistration>,
+                                          userKey: string) => {
+            var eventRegistration: ClientEventRegistration = registrationsOnUserKey.get(connection);
+            if (eventRegistration !== undefined) {
+                this.client.getInvocationService().removeEventHandler(eventRegistration.correlationId.toNumber());
+            }
+        });
+    }
+
+    invokeRegistrationFromRecord(userRegistrationKey: string, connection: ClientConnection): Promise<ClientEventRegistration> {
```
Comment (asimarslan) [inline src/invocation/InvocationService.ts]: is this method share any code with `registerListenerInternal` ?  A shared invoke method will increase readability, if possible.

File: src/proxy/MapProxy.ts [hunk 3]
```diff
@@ -552,32 +552,42 @@ export class MapProxy<K, V> extends BaseProxy implements IMap<K, V> {
...
         var request: ClientMessage;
+        var encodeFunc: Function;
         var handler: Function;
         var responser: Function;
         if (key && predicate) {
             var keyData = this.toData(key);
             var predicateData = this.toData(predicate);
-            request = MapAddEntryListenerToKeyWithPredicateCodec.encodeRequest(this.name, keyData,
-                predicateData, includeValue, flags, false);
+            encodeFunc = (localOnly: boolean) => {
```
Comment (asimarslan) [inline src/proxy/MapProxy.ts]: Doesn't the encodeFunc always create the same message? so why not just use that encoded clientMessage instead and save some cpu cycles ?

File: src/invocation/InvocationService.ts (hunk 2 above)
Comment (mustafaiman) [reply src/invocation/InvocationService.ts]: This method takes the record of a previously registered listener and invokes the registration again when a connection is restored whereas `registerListenerInternal` registers a register for the first time. `registerListenerInternal` creates records of listeners while this method uses those records. Invocation part is only `this.client.getInvocationService().invoke(invocation)` part. The actions taken at the result of invocation are also different. It is not worth extracting another method and sharing it.

File: src/invocation/RegistrationKey.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
+export class RegistrationKey {
+    private readonly userRegistrationId: string;
+    private handler: Function;
+    private encoder: Function;
```
Comment (asimarslan) [inline src/invocation/RegistrationKey.ts]: Can we rename all encoder and decoder naming to give the meaning of

* `registration Encoder Function` ( something like `registerEncodeFunc`) 
* `registration Decoder Function`
* `deregistration Encoder Function`
* `deregistration Decoder Function`

I can see encode decode keywords everywhere, it's hard to follow.

Comment (mustafaiman) [reply src/invocation/RegistrationKey.ts]: did that.

File: src/proxy/MapProxy.ts (hunk 3 above)
Comment (mustafaiman) [reply src/proxy/MapProxy.ts]: I think it can be done that way. We actually needed to pass encode function to listenerservice for only de-registration because listener-id that user sees and actual listener-id are different. So I made both registration and de-registration apis similar.

**Comment by mustafaiman** [comment]
@asimarslan I changed `registerListener` api so it takes `ClientMessage` now instead of register listener encode function.

**Comment by mustafaiman** [comment]
verify

File: src/ListenerService.ts [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+            }).then(() => {
+                this.activeRegistrations.delete(userRegistrationKey);
+                this.userRegistrationKeyInformation.delete(userRegistrationKey);
+            });
+        });
+
+        return deferred.promise;
+    }
+
+    private trySyncConnectToAllConnections(): Promise<void> {
```
Comment (asimarslan) [inline src/ListenerService.ts]: This method should timeout after invocation Timeout period.
